src/service-override/storage.ts

```diff
@@ -0,0 +1,144 @@
+import { Emitter, Event } from 'vs/base/common/event'
+import { IStorage, IStorageDatabase, IStorageItemsChangeEvent, IUpdateRequest, Storage } from 'vs/base/parts/storage/common/storage'
+import { IEditorOverrideServices } from 'vs/editor/standalone/browser/standaloneServices'
+import { SyncDescriptor } from 'vs/platform/instantiation/common/descriptors'
+import { AbstractStorageService, IStorageService, StorageScope as VSStorageScope } from 'vs/platform/storage/common/storage'
+import { IUserDataProfile } from 'vs/platform/userDataProfile/common/userDataProfile'
+import { IAnyWorkspaceIdentifier } from 'vs/platform/workspace/common/workspace'
+
+export enum StorageScope {
+  APPLICATION = 'application',
+  PROFILE = 'profile',
+  WORKSPACE = 'user',
+}
+
+export interface IStorageProvider {
+  read (scope: StorageScope): Map<string, string> | undefined
+  write (scope: StorageScope, data: Map<string, string>): Promise<void>
+  close? (scope: StorageScope): Promise<void>
+  onDidChange? (callback: (event: IStorageItemsChangeEvent) => void): void
+}
+
+class ExternalStorage extends Storage {
+  constructor (
+    scope: StorageScope,
+    provider: IStorageProvider
+  ) {
+    const items = provider.read(scope)
+
+    super(new ExternalStorageDatabase(scope, provider, items))
+
+    if (items != null) {
+      for (const [key, value] of items) {
+        this.items.set(key, value)
+      }
+    }
+  }
+}
+
+class ExternalStorageDatabase implements IStorageDatabase {
+  readonly onDidChangeItemsExternal: Event<IStorageItemsChangeEvent>
+
+  constructor (
+    private readonly scope: StorageScope,
+    private readonly provider: IStorageProvider,
+    private readonly items = new Map<string, string>()
+  ) {
+    if (this.provider.onDidChange != null) {
+      const emitter = new Emitter<IStorageItemsChangeEvent>()
+
+      this.provider.onDidChange((data: IStorageItemsChangeEvent) => emitter.fire(data))
+
+      this.onDidChangeItemsExternal = emitter.event
+    } else {
+      this.onDidChangeItemsExternal = Event.None
+    }
```

Comment:
Isn't it equivalent to `this.onDidChangeItemsExternal = this.provider.onDidChange ?? Event.None` ?

Reply:
Ultimately yes, that's the equivalent, however signatures do not match, but if you reckon it's better for `provider.onDidChange` to match the `Event` signature instead - will change to that!

Reply:
> I'm not really sure to understand what you mean by
> 
> > signatures do not match

Well the difference is that `provider.onDidChange` return type is `void` and `Event`'s return type is `IDisposable`. I can either change provider to use `IDisposable` or keep the current solution.

Reasoning was that forcing implementer to return disposable is a bit too much as it most likely won't be called at all.